Products/GenericSetup/tool.py

```diff
@@ -1166,7 +1178,16 @@ def _runImportStepsFromContext(self,
 
         detect_steps = steps is None
 
-        for profile_id in chain:
+        last_index = len(chain) - 1
+        for index, profile_id in enumerate(chain):
+            if not always_apply_profiles and index != last_index:
+                # Check if dependency profile was already applied.
+                if self.getLastVersionForProfile(profile_id) != 'unknown':
+                    # Profile was already applied.
+                    # Maybe apply upgrade steps, if any, otherwise continue.
+                    if upgrade_dependencies:
+                        self.upgradeProfile(profile_id)
+                    continue
             context = self._getImportContext(profile_id, purge_old, archive)
             self.applyContext(context)
```

Comment:
Thanks, the logic is still a bit weird but makes more sense now. This is the sort of method where I'd normally spend time working out how to reduce the number of branches in it but sometimes life is too short.